marketing.js

```diff
@@ -3,113 +3,65 @@
-const AWS_S3_BUCKET = '*** TODO ***'
-const AWS_S3_PREFIX = '*** TODO ***'
-
 const MAX_EVENTS_PER_BATCH = 10
 const HMAC_KEY = process.env.FXA_AMPLITUDE_HMAC_KEY
 const API_KEY = process.env.FXA_AMPLITUDE_API_KEY
-
-if (process.argv.length !== 3) {
-  console.error(`Usage: ${process.argv[1]} {YYYY-MM-DD-hh-mm | LOCAL PATH}`)
-  console.error('If specifying YYYY-MM-DD-hh-mm as the arg, note that the script will try to send events')
-  console.error('for all times from YYYY-MM-DD-hh-mm to the most recent available in S3. If any times in')
-  console.error('that range are missing, they will be skipped without failing the process.')
-  process.exit(1)
-}
+const WORKERS = process.env.FXA_AMPLITUDE_WORKERS || 8
 
 if (! HMAC_KEY || ! API_KEY) {
   console.error('Error: You must set FXA_AMPLITUDE_HMAC_KEY and FXA_AMPLITUDE_API_KEY environment variables')
   process.exit(1)
 }
 
-processData()
-  .catch(error => {
-    console.error(error.stack)
-    process.exit(1)
-  })
-
-function processData () {
-  const timeParts = TIMESTAMP.exec(process.argv[2])
-
-  if (timeParts && timeParts.length === 6) {
-    if (! AWS_ACCESS_KEY || ! AWS_SECRET_KEY) {
-      console.error('Error: You must set AWS_ACCESS_KEY and AWS_SECRET_KEY environment variables')
-      process.exit(1)
-    }
-
-    return processDataFromS3(timeParts.slice(1))
-      .then(counts => counts.reduce((sum, count) => sum + count, 0))
-  }
-
-  return processStream(fs.createReadStream(process.argv[2]))
-}
-
-function processStream (stream) {
+module.exports.processStream = function processStream (stream) {
   let eventCount = 0, batch = [], error
 
+  let cargo = async.cargo(async (tasks, callback) => {
+    await send(tasks)
+    callback()
+  }, MAX_EVENTS_PER_BATCH)
+  cargo.concurrency = WORKERS
+
   process.on('exit', () => {
     console.log(`Sent ${eventCount} events!`)
   })
 
   return new Promise((resolve, reject) => {
+    cargo.drain = () => {
+      resolve(eventCount)
+    }
+
     stream
       .pipe(new AutoDetectDecoderStream())
       .pipe(csv())
-      .on('data', row => {
+      .on('data', async (row) => {
         const event = createEvent(row)
         if (! event) {
           return
         }
 
-        batch.push(event)
-
-        if (batch.length < MAX_EVENTS_PER_BATCH) {
-          return
-        }
-
-        const localBatch = batch.slice()
-        batch = []
-        send(localBatch).catch(reject)
-      })
-      .on('end', () => {
-        if (error) {
-          reject(error)
-        } else if (batch.length === 0) {
-          resolve(eventCount)
-        } else {
-          send(batch)
-            .then(() => resolve(eventCount))
-            .catch(reject)
-        }
+        cargo.push(event)
       })
   })
 
-  function send (localBatch) {
-    if (error) {
-      return Promise.reject(error)
+  async function send (localBatch) {
+    const body = await sendBatch(localBatch)
+
+    if (body == "success") {
```

Comment:
Personally this feels like coupling too tightly to an implementation detail of the Amplitude API. If they change the response body at any point we're going to have to fix the script and re-send a bunch of failed events. I thought the previous error-handling strategy was more robust.

But I'm not the one maintaining it, so no biggie if this is preferable. Just wanted to call it out.

Reply:
> fix the script and re-send a bunch of failed events

Well, not the re-send part, the events will have been sent successfully of course.